### PR TITLE
feat: Unify traces of sub-pipelines within pipelines with Langfuse

### LINKED ISSUE #1604: `AsyncPipeline` Creates Multiple Traces Instead of a Single Unified Trace in Langfuse
**Describe the Bug:**  
When using the `AsyncPipeline` instead of the synchronous `Pipeline`, multiple trace entries are created in Langfuse for what should be a **single execution** of the pipeline. This behavior makes it difficult to get a unified view of the request flow.

Additionally, the following warning is logged during execution:

```
WARNING: 2025-04-03,17:49:55: tracer.py:243: Creating a new trace without a parent span is not recommended for operation 'haystack.async_pipeline.run'.
```

This suggests that the trace context is not being properly propagated across async components or pipeline runs.

---

**Code to Reproduce:**

```python
from haystack import AsyncPipeline
from haystack_integrations.components.tracers.langfuse import LangfuseConnector

pipeline = AsyncPipeline()

# Add components (simplified example)
pipeline.add_component("tracer", LangfuseConnector(name="Async Pipeline"))
pipeline.add_component("retriever", SomeRetrieverComponent())
pipeline.add_component("generator", SomeGeneratorComponent())

# Define connections
pipeline.connect("retriever", "generator")

# Run the async pipeline
await pipeline.run({"retriever": {"query": "What is Haystack?"}})
```

---

**Observed Behavior:**  
- Multiple traces appear in Langfuse for a single pipeline run.
- The LangfuseConnector logs a warning about missing parent span context.
- Lack of unified tracing breaks end-to-end visibility for async workflows.

---

**Expected Behavior:**  
A **single trace** should be created and propagated throughout the entire async pipeline execution, just like in the synchronous `Pipeline`. This would ensure consistent observability and debugging in Langfuse.

Image

**Describe your environment (please complete the following information):**
 - OS: Linux, WSL
 - haystack-ai==2.12.0
 - langfuse-haystack==0.8.0

### LINKED ISSUE #1605: Support Unified Tracing with `LangfuseConnector` Across Main and Sub-Pipelines in Haystack
Please provide guidance or documentation on how to configure Haystack pipelines with `LangfuseConnector` to ensure a **single, unified trace** across both main pipelines and sub-pipelines. If this functionality is not currently supported, I would like to propose it as a **feature request** to enhance traceability in complex pipeline setups.

---

**Description:**  
When using Haystack pipelines that include sub-pipelines alongside the `LangfuseConnector`, it's currently unclear how to maintain a **consistent trace context** throughout the entire execution flow.

At present, it appears that each pipeline (main and sub-pipelines) may generate **separate traces**, which makes it difficult to monitor or debug the full journey of a request from the top-level pipeline into its nested components in Langfuse.

---

**Reproducible Example:**  

Here's a simplified example to demonstrate the issue. We have a main pipeline and two sub-pipelines. A `LangfuseConnector` is included in the main pipeline for tracing purposes.

```python
from haystack import Pipeline
...

# Sub-pipeline 1: Embedding and Retrieval
@component
class RetrievalPipeline:
    def run():
        pipeline = Pipeline()
        document_store = InMemoryDocumentStore(embedding_similarity_function="cosine")
        text_embedder = SentenceTransformersTextEmbedder()
        retriever = InMemoryEmbeddingRetriever(document_store=document_store)

        pipeline.add_component("text_embedder", text_embedder)
        pipeline.add_component("retriever", retriever)
        pipeline.connect("text_embedder.embedding", "retriever.query_embedding")

        return pipeline.run()

# Sub-pipeline 2: Answer Generation
@component
class GenerationPipeline:
    def run():
        pipeline = Pipeline()
        generator = OpenAIGenerator(api_key="YOUR_API_KEY")  # Replace with actual key or use env var
        pipeline.add_component("generator", generator)

        return pipeline.run()

# Main pipeline
main_pipeline = Pipeline()

langfuse_connector = Langfuse()
retrieval_pipeline = RetrievalPipeline()
generation_pipeline = GenerationPipeline()

main_pipeline.add_component("langfuse", langfuse_connector)
main_pipeline.add_component("retrieval", retrieval_pipeline)
main_pipeline.add_component("generation", generation_pipeline)

main_pipeline.connect("retrieval", "generation")

query = "What is Haystack?"
main_pipeline.invoke()
```

---

**Expected Behavior:**  
We would expect a **single trace** in Langfuse that includes the full execution of the main pipeline along with its sub-pipelines (retrieval and generation steps). This would provide an end-to-end view of the request lifecycle in a unified format.

---

**Additional Notes:**  
It seems there is a `span_handler` parameter available in the `LangfuseConnector`, and we wonder if customizing it could allow propagation of trace context across sub-pipelines. If there are any recommended approaches or workarounds using this, documentation or examples would be greatly appreciated.

---

**Feature Request (if not currently supported):**  
If unified trace context propagation is not currently supported out of the box, please consider adding this functionality. It would significantly improve the developer experience and observability for users building modular, nested pipelines in Haystack.

---

### PR DESCRIPTION
**Pull request by sjrl**
### Related Issues

- fixes #1604 and #1605 

### Proposed Changes:

 
 
- Correctly pass the parent_span in the situation we have a sub-pipeline within a pipeline

### How did you test it?

- Added integration test and tested locally
- Also expanded unit tests and did some refactoring

### Notes for the reviewer

Here is some example code

```python
import os

os.environ["OPENAI_API_KEY"] = ""
os.environ["LANGFUSE_SECRET_KEY"] = ""
os.environ["LANGFUSE_PUBLIC_KEY"] = ""
os.environ["HAYSTACK_CONTENT_TRACING_ENABLED"] = "true"

from haystack import Pipeline, SuperComponent
from haystack.components.builders import ChatPromptBuilder
from haystack.components.generators.chat import OpenAIChatGenerator
from haystack.dataclasses import ChatMessage
from haystack.utils.auth import Secret

from haystack_integrations.components.connectors.langfuse import LangfuseConnector


sub_pipe = Pipeline()
sub_pipe.add_component("llm", OpenAIChatGenerator())
super_llm = SuperComponent(
    pipeline=sub_pipe,
)

top_sub_pipe = Pipeline()
top_sub_pipe.add_component("prompt_builder", ChatPromptBuilder())
top_sub_pipe.add_component("super_llm", super_llm)
top_sub_pipe.connect("prompt_builder.prompt", "super_llm.messages")
super_prompt_and_llm = SuperComponent(
    pipeline=top_sub_pipe,
)

# Main pipeline
pipe = Pipeline()
pipe.add_component("tracer", LangfuseConnector("Sub-Sub Pipeline Test"))
pipe.add_component("super_prompt_and_llm", super_prompt_and_llm)

messages = [
    ChatMessage.from_system("Always respond in German even if some input data is in other languages."),
    ChatMessage.from_user("Tell me about {{location}}"),
]

response = pipe.run(
    data={
        "super_prompt_and_llm": {
            "template_variables": {"location": "Berlin"},
            "template": messages,
        },
        "tracer": {
            "invocation_context": {"some_key": "some_value"},
        },
    }
)
print(response["super_prompt_and_llm"]["replies"][0])
print(response["tracer"]["trace_url"])
```

which produces this trace in Langfuse

Screenshot 2025-04-08 at 11 49 28

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/langfuse/tests/test_langfuse_connector.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        langfuse_connector = LangfuseConnector.from_dict(data)
+        assert langfuse_connector.name == "Chat example - OpenAI"
+        assert langfuse_connector.public is True
+        assert langfuse_connector.secret_key == Secret.from_env_var("LANGFUSE_SECRET_KEY")
+        assert langfuse_connector.public_key == Secret.from_env_var("LANGFUSE_PUBLIC_KEY")
+        assert isinstance(langfuse_connector.span_handler, CustomSpanHandler)
+
+    def test_pipeline_serialization(self, monkeypatch):
```
Comment (sjrl) [inline integrations/langfuse/tests/test_langfuse_connector.py:176]: This was moved from the `test_tracing.py` file

File: integrations/langfuse/tests/test_langfuse_span.py [hunk 2]
```diff
@@ -1,71 +0,0 @@
-# SPDX-FileCopyrightText: 2023-present deepset GmbH <info@deepset.ai>
```
Comment (sjrl) [inline integrations/langfuse/tests/test_langfuse_span.py:1]: This file got refactored into the `test_tracer.py` file, following our standard practice of having the test files follow the same name structure as our module files.

File: integrations/langfuse/tests/test_tracing.py [hunk 3]
```diff
@@ -107,114 +102,80 @@ def test_tracing_integration(llm_class, env_var, expected_trace, pipeline_fixtur
...
-            elif len(reply.text) > 10:
-                span._span.update(level="WARNING", status_message="Response too long (> 10 chars)")
-            else:
-                span._span.update(level="DEFAULT", status_message="Success")
-
-
 @pytest.mark.integration
-def test_custom_span_handler():
-    """Test that custom span handler properly sets Langfuse levels."""
+def test_tracing_with_sub_pipelines():
```
Comment (sjrl) [inline integrations/langfuse/tests/test_tracing.py:106]: This is the new test, that checks that when using sub-pipelines within a pipeline we correctly collect the traces together.

File: integrations/langfuse/src/haystack_integrations/tracing/langfuse/tracer.py [hunk 4]
```diff
@@ -353,19 +357,22 @@ def trace(
...
-            )
+        # Create a new span context
+        span_context = SpanContext(
+            name=span_name,
+            operation_name=operation_name,
+            component_type=component_type,
+            tags=tags,
+            # We use the current active span as the parent span if not provided to handle nested pipelines
+            # The nested pipeline (or sub-pipeline) will be a child of the current active span
+            parent_span=parent_span or self.current_span(),
```
Comment (sjrl) [inline integrations/langfuse/src/haystack_integrations/tracing/langfuse/tracer.py:368]: This line is what fixed the issue

**Comment by sjrl** [comment]
pinging @vblagoje in case you are interested

File: integrations/langfuse/tests/test_langfuse_connector.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        assert tracer_params["secret_key"]["env_vars"] == ["LANGFUSE_SECRET_KEY"]
+        assert isinstance(tracer_params["public_key"], dict)
+        assert tracer_params["public_key"]["type"] == "env_var"
+        assert tracer_params["public_key"]["env_vars"] == ["LANGFUSE_PUBLIC_KEY"]
+
+        # Deserialize
+        new_pipe = Pipeline.from_dict(serialized)
+
+        # Verify pipeline is the same
+        assert new_pipe == pipe
```
Comment (Amnah199) [inline integrations/langfuse/tests/test_langfuse_connector.py:213]: Great to see that you've introduced multiple test cases. Maybe in future we can cover some edge cases`test_invalid_span_handler` to verify how the system behaves. But its your call.

**Review by Amnah199 — APPROVED** [review]
Great work! Although I may not have full context on Langfuse, I reviewed the code based on my understanding. I am guessing @vblagoje has also had a chance to take a look.

File: integrations/langfuse/tests/test_langfuse_connector.py (hunk 5 above)
Comment (sjrl) [reply integrations/langfuse/tests/test_langfuse_connector.py:213]: Good point! I think that is something we can do in the future.
